### PR TITLE
Kernel: Add support for SD host controllers on the PCI bus

### PR DESCRIPTION
**Pull request by mrkct**
These commits allow using SD cards on x86 too + a fix for a kernel page fault when unmounting a disk.

Note that there's still no support for hot-plugging cards: only cards connected since boot will work.

To test this:

**Create an image for the SD card**
```
touch sd-card.img
qemu-img resize sd-card.img 2G
mkfs.ext2 sd-card.img
mkdir mountpoint
sudo mount sd-card.img mountpoint
sudo sh -c 'echo Hello world from the SD card > mountpoint/Hello.txt'
sudo umount mountpoint
```

**Run with the SD card inserted**
```
SERENITY_EXTRA_QEMU_ARGS="-device sdhci-pci -device sd-card,drive=mydrive -drive id=mydrive,if=none,format=raw,file=$(pwd)/sd-card.img" Meta/serenity.sh run
```

**Inside Serenity**
```
su
mkdir sd-card
mount /dev/hdb sd-card
ls sd-card
```

### CODE REVIEW

File: Kernel/FileSystem/BlockBasedFileSystem.cpp [hunk 1]
```diff
@@ -101,11 +101,11 @@ class DiskCache {
 
 private:
     mutable NonnullRefPtr<BlockBasedFileSystem> m_fs;
+    NonnullOwnPtr<KBuffer> m_cached_block_data;
+    NonnullOwnPtr<KBuffer> m_entries;
```
Comment (Hendiadyoin1) [inline Kernel/FileSystem/BlockBasedFileSystem.cpp:105]: Maybe add a comment, why order matters here

File: Kernel/Storage/SD/PCISDHostController.cpp [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    if (slot_information_register.slots_available() != 1) {
+        // TODO: Support multiple slots
+        dmesgln("SD Host Controller has {} slots, but we currently only support using only one", slot_information_register.slots_available());
+    }
+
+    auto physical_address_of_sdhc_registers = PhysicalAddress {
+        PCI::get_BAR(device_identifier, static_cast<PCI::HeaderType0BaseRegister>(slot_information_register.first_bar_number))
+    };
+    if (auto maybe_registers = Memory::map_typed_writable<SD::HostControlRegisterMap volatile>(physical_address_of_sdhc_registers); !maybe_registers.is_error())
+        m_registers = maybe_registers.release_value();
```
Comment (Hendiadyoin1) [inline Kernel/Storage/SD/PCISDHostController.cpp:35]: looks like a job for `release_value_but_fixme_should_propagate_errors`
or is no registers a valid state?

File: Kernel/Storage/SD/PCISDHostController.cpp [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+    if (slot_information_register.slots_available() != 1) {
+        // TODO: Support multiple slots
+        dmesgln("SD Host Controller has {} slots, but we currently only support using only one", slot_information_register.slots_available());
+    }
+
+    auto physical_address_of_sdhc_registers = PhysicalAddress {
+        PCI::get_BAR(device_identifier, static_cast<PCI::HeaderType0BaseRegister>(slot_information_register.first_bar_number))
+    };
+    if (auto maybe_registers = Memory::map_typed_writable<SD::HostControlRegisterMap volatile>(physical_address_of_sdhc_registers); !maybe_registers.is_error())
```
Comment (ADKaster) [inline Kernel/Storage/SD/PCISDHostController.cpp:34]: If we don't actually properly map this, isn't the device in an invalid state? How do we detect that and not attach the device after all? Should this be done in the factory function or in an overridden initialize() method instead?

File: Kernel/FileSystem/BlockBasedFileSystem.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (Hendiadyoin1) [reply Kernel/FileSystem/BlockBasedFileSystem.cpp:105]: :ninja:'d

**Comment by ADKaster** [comment]
Can you add a commit with an option+environment variable in Meta/run.sh (e.g. `SERENITY_USE_SDCARD` or such, similar to the NVME one) that uses an SD card for the main rootfs?

**Comment by mrkct** [comment]
Note that booting from the SD card is extremely slow right now(we're talking 30+ minutes boot times here).

It's here so that others can also find it and maybe improve it.

**Review by supercomputer7 — APPROVED** [review]
Seems OK to me :)

**Review by ADKaster — APPROVED** [review]

**Pull request merged**